Spigot-Server-Patches/0333-Add-entity-count-cache.patch

```diff
@@ -0,0 +1,148 @@
+
+
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 004c3ec4..09c2bc32 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -50,10 +50,41 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+     // Spigot start - guard entity list from removals
+     public final List<Entity> entityList = new java.util.ArrayList<Entity>()
+     {
++        // Paper start - entity count cache
++        @Override
++        public boolean addAll(Collection<? extends Entity> c) {
++            for(Entity e : c) {
++                updateEntityCount(e, true);
++            }
++
++            return super.addAll(c);
++        }
++
++        @Override
++        public boolean removeAll(Collection<?> c) {
++            for(Object e : c) {
++                if(e instanceof Entity) {
++                    updateEntityCount((Entity)e, false);
++                }
++            }
++
++            return super.removeAll(c);
++        }
++
++        @Override
++        public boolean add(Entity e) {
++            updateEntityCount(e, true);
++
++            return super.add(e);
++        }
++
++        // Paper end
++
+         @Override
+         public Entity remove(int index)
+         {
+             guard();
++            updateEntityCount(get(index), false); // Paper
+             return super.remove( index );
+         }
+ 
+@@ -61,6 +92,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         public boolean remove(Object o)
+         {
+             guard();
++            if(o instanceof Entity) updateEntityCount((Entity)o, false); // Paper
+             return super.remove( o );
+         }
+ 
+@@ -82,6 +114,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+     public final Map<String, EntityHuman> playersByName = Maps.newHashMap(); // Paper - World EntityHuman Lookup Optimizations
+     public final List<Entity> k = Lists.newArrayList();
+     protected final IntHashMap<Entity> entitiesById = new IntHashMap();
++    private Map<Class<?>, Integer> countCache = new HashMap(); // Paper
+     private final long G = 16777215L;
+     private int H; public int getSkylightSubtracted() { return this.H; } public void setSkylightSubtracted(int value) { this.H = value;} // Paper - OBFHELPER
+     protected int m = (new Random()).nextInt();
+@@ -2439,19 +2472,16 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+ 
+     }
+ 
+-    public int a(Class<?> oclass) {
++    // Paper start - entity count cache
++    private int countEntityType(Class<?> oclass) {
+         int i = 0;
+         Iterator iterator = this.entityList.iterator();
+ 
+         while (iterator.hasNext()) {
+             Entity entity = (Entity) iterator.next();
+-            if (entity.shouldBeRemoved) continue; // Paper
```

Comment:
so i actually removed this line because the count is being updated by the actual modifications to *World#entityList*, if i didnt could entities that are marked shouldBeRemoved, the count could end up off (at the end of the tick when its time to actually remove them from the list).

Reply:
Then something has to be refactored here.
This boolean represents "This entity should not of been in this list in the first place"

You'll need to hook into where this is set to true/false at (at removeEntity, and the world add method) and hook count changes there too.